Add microscopy tutorial (#1036)
ivadomed/ivadomed: docs/source/tutorials/two_classes_microscopy_seg_2d_unet.rst

> Two-classes microscopy segmentation with 2D U-Net

Reviewer: ```suggestion
Two-class microscopy segmentation with 2D U-Net
```

> =================================================

Reviewer: ```suggestion
===============================================
```

In this tutorial we will learn the following features:

> - Training of a segmentation model (U-Net 2D) with a two classes labels on a single contrast on microscopy PNG images,

Reviewer: ```suggestion
- Training of a segmentation model (U-Net 2D) with two-class labels on a single contrast on microscopy PNG images,
```

- Testing of a trained model and computation of evaluation metrics,

- Visualization of the outputs of a trained model.

Download dataset
----------------

> We will use a publicly-available dataset consisting of 10 microscopy samples of rat spinal cord.

Reviewer: ```suggestion
We will use a publicly available dataset consisting of 10 microscopy samples of rat spinal cord.
```

> We will use a publicly-available dataset consisting of 10 microscopy sample of rat spinal cord.

Reviewer: ```suggestion
We will use a publicly-available dataset consisting of 10 microscopy samples of rat spinal cord.
```

the ``ivadomed/config/`` and the documentation is available in :doc:`../configuration_file`.

In this tutorial we will use the configuration file: ``ivadomed/config/config_microscopy.json``.
First off, copy this configuration file in your local directory (to avoid modifying the source file):

.. code-block:: bash

   cp <PATH_TO_IVADOMED>/ivadomed/config/config_microscopy.json .

Then, open it with a text editor. Below we will discuss some of the key parameters to perform a two-class 2D
microscopy segmentation training.

- ``command``: Action to perform. Here, we want to train a model, so we set the fields as follows:

  .. code-block:: xml

     "command": "train"

- ``path_output``: Folder name that will contain the output files (e.g., trained model, predictions, results).

  .. code-block:: xml

     "path_output": "log_microscopy_sem"

- ``loader_parameters:path_data``: Location of the dataset. As discussed in `Data <../data.html>`__, the dataset
  should conform to the BIDS standard. Modify the path so it points to the location of the downloaded dataset.

  .. code-block:: xml

     "path_data": "data_axondeepseg_sem"

> - ``loader_parameters:bids_config``: Location of the custom bids configuration file required for microscopy

Reviewer: ```suggestion
- ``loader_parameters:bids_config``: Location of the custom BIDS configuration file required for microscopy
```

  file indexing.

  .. code-block:: xml

     "bids_config": "ivadomed/config/config_bids.json",

> - ``loader_parameters:target_suffix``: Suffix of the ground truths segmentation. The ground truths are located

Reviewer: ```suggestion
- ``loader_parameters:target_suffix``: Suffix of the ground truth segmentations. The ground truths are located
```

  train/validation/test sets.
- ``split_dataset:train_fraction``: Fraction of the dataset's ``sample_id`` in the train set. In our case ``0.6``.
- ``split_dataset:test_fraction``: Fraction of the dataset's ``sample_id`` in the test set. In our case ``0.1``.

  .. code-block:: xml

      "split_method" : "sample_id"
      "train_fraction": 0.6
      "test_fraction": 0.1

- ``training_time:num_epochs``: the maximum number of epochs that will be run during training. Each epoch is composed
  of a training part and an evaluation part. It should be a strictly positive integer. In our case, we will use
  50 epochs.

  .. code-block:: xml

     "num_epochs": 50

- ``default_model:length_2D``: Size of the 2D patches used as model’s input tensors. We recommend using patches
  between 256x256 and 512x512. In our case, we use patches of 256x256.
- ``default_model:stride_2D``: Pixels’ shift over the input matrix to create 2D patches. In our case, we use
  a stride of 244 pixels in both dimensions, resulting in an overlap of 12 pixels between patches.

  .. code-block:: xml

     "length_2D": [256, 256]
     "stride_2D": [244, 244]

> - ``postprocessing:binarize_maxpooling``: Use to binarize predictions across all classes in multiclass models.

Reviewer: ```suggestion
- ``postprocessing:binarize_maxpooling``: Used to binarize predictions across all classes in multiclass models.
```

  .. code-block:: xml

      "binarize_maxpooling": {}

> - ``transformation:Resample``: Use to resample images to a common resolution (in mm) before splitting into patches,

Reviewer: ```suggestion
- ``transformation:Resample``: Used to resample images to a common resolution (in mm) before splitting into patches,
```

   Creating model directory: log_microscopy_sem/model_seg_rat_axon-myelin_sem

   Initialising model's weights from scratch.

   Scheduler parameters: {'name': 'CosineAnnealingLR', 'base_lr': 1e-05, 'max_lr': 0.01}

   Selected Loss: DiceLoss
   with the parameters: []
   Epoch 1 training loss: -0.6894.
   Epoch 1 validation loss: -0.7908.

After 50 epochs (see ``"num_epochs"`` in the configuration file), the Dice score on the validation set should be ~85%.

.. note::

   When loading the images for training or evaluation, a temporary NIfTI file will be created for each images in the
   dataset directory (``path_data``) alongside the original PNG files.

Evaluate model
--------------

To test the trained model on the testing sub-dataset and compute evaluation metrics, run:

.. code-block:: bash

   ivadomed -c config_microscopy.json --test

If you prefer to use config files over CLI flags, set "command" to the following in you config file:

.. code-block:: xml

   "command": "test"

Then run:

.. code-block:: bash

   ivadomed -c config_microscopy.json

The model's parameters will be displayed in the terminal, followed by a preview of the results for each image.

> The resulting segmentations are saved for each image in the ``<PATH_TO_OUT_DIR>/pred_masks`` while a csv file,

Reviewer: ```suggestion
The resulting segmentations are saved for each image in the ``<PATH_TO_OUT_DIR>/pred_masks`` while a CSV file,
```
